bascule/key/cache.go

```diff
@@ -0,0 +1,219 @@
+package key
+
+import (
+	"github.com/Comcast/webpa-common/concurrent"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	// dummyKeyId is used when no actual keyID is necessary.
+	dummyKeyId = ""
+)
+
+// Cache is a Resolver type which provides caching for keys based on keyID.
+//
+// All implementations will block the first time a particular key is accessed
+// and will initialize the value for that key.  Thereafter, all updates happen
+// in a separate goroutine.  This allows HTTP transactions to avoid paying
+// the cost of loading a key after the initial fetch.
+type Cache interface {
+	Resolver
+
+	// UpdateKeys updates all keys known to this cache.  This method makes
+	// a best-effort to avoid blocking other goroutines which use ResolveKey,
+	// which may mean copy-on-write semantics.
+	//
+	// The first return value is the count of keys for which attempts were
+	// made to update.
+	//
+	// UpdateKeys may run multiple I/O operations.  The second return value is a slice of
+	// errors that occurred while it attempted to update each key.  Exactly one (1)
+	// attempt will be made to update each key present in the cache, regardless
+	// of any update errors for each individual key.  This slice may be nil if no
+	// errors occurred.
+	UpdateKeys() (int, []error)
+}
+
+// basicCache contains the internal members common to all cache implementations
+type basicCache struct {
```

Comment:
The caching bit needs some refactoring.  This approach, with nested types for reuse, doesn't work very well for unmarshalling or dependency injection.